runtime/runc_jailer.go

```diff
@@ -0,0 +1,370 @@
+}
+
+func (j runcJailer) JailPath() vm.Dir {
+	return vm.Dir(j.RootPath())
+}
+
+func (j *runcJailer) BuildJailedMachine(cfg *Config, machineConfig *firecracker.Config, vmID string) ([]firecracker.Opt, error) {
+	handler := j.BuildHandler(cfg, &machineConfig.SocketPath, vmID)
+	fifoHandler := j.BuildLinkFifoHandler()
+	client := firecracker.NewClient(machineConfig.SocketPath, j.logger, machineConfig.Debug)
+
+	opts := []firecracker.Opt{
+		firecracker.WithProcessRunner(j.jailerCommand(vmID)),
+		firecracker.WithClient(client),
+		func(m *firecracker.Machine) {
+			m.Handlers.FcInit = m.Handlers.FcInit.Prepend(handler)
+			m.Handlers.FcInit = m.Handlers.FcInit.AppendAfter(firecracker.CreateLogFilesHandlerName, fifoHandler)
+		},
+	}
+
+	return opts, nil
+}
+
+// BuildHandler will populate the jail with the necessary files, which may be
+// device nodes, hard links, and/or bind-mount targets
+func (j *runcJailer) BuildHandler(cfg *Config, socketPath *string, vmID string) firecracker.Handler {
+	ociBundlePath := j.OCIBundlePath()
+	rootPath := j.RootPath()
+	*socketPath = filepath.Join(rootPath, "api.socket")
+
+	return firecracker.Handler{
+		Name: jailerHandlerName,
+		Fn: func(ctx context.Context, m *firecracker.Machine) error {
+			const mode = os.FileMode(0700)
+			// Create the proper paths needed for the runc jailer
+			j.logger.WithField("rootPath", rootPath).Debug("Creating root drive path")
+			if err := mkdirAndChown(rootPath, mode, j.uid, j.gid); err != nil {
+				return errors.Wrapf(err, "%s failed to mkdirAndChown", rootPath)
+			}
+
+			rootPathToConfig := filepath.Join(ociBundlePath, "config.json")
+			j.logger.WithField("rootPathToConfig", rootPathToConfig).Debug("Copying config")
+			if err := copyFile(runcConfigPath, rootPathToConfig, 0444); err != nil {
+				return errors.Wrapf(err, "failed to copy config from %v to %v", runcConfigPath, rootPathToConfig)
+			}
+
+			j.logger.Debug("Overwritting process args of config")
+			if err := j.overwriteConfig(cfg, filepath.Base(m.Cfg.SocketPath), rootPathToConfig); err != nil {
+				return errors.Wrap(err, "failed to overwrite config.json")
+			}
+
+			// copy the firecracker binary
+			j.logger.WithField("root path", rootPath).Debug("copying firecracker binary")
+			if err := copyFile(cfg.FirecrackerBinaryPath, rootPath, 0700); err != nil {
+				return errors.Wrapf(err, "could not copy firecracker binary from path %v", cfg.FirecrackerBinaryPath)
+			}
+			if err := os.Chown(
+				filepath.Join(rootPath, filepath.Base(cfg.FirecrackerBinaryPath)),
+				int(j.uid),
+				int(j.gid),
+			); err != nil {
+				return errors.Wrap(err, "failed to change ownership of binary")
+			}
+
+			// Link the kernel image
+			newKernelImagePath := filepath.Join(rootPath, kernelImageFileName)
+			j.logger.WithField("newKernelImagePath", newKernelImagePath).Debug("linking kernel image")
+			if err := os.Link(m.Cfg.KernelImagePath, newKernelImagePath); err != nil {
```

Comment:
Hard links only work across the same filesystem.  Do we want to enforce that the kernel image and the jail directory are on the same filesystem?

I think it'd make sense to start with copying for now, and then explore read-only bindmounts if we want to optimize later.

Reply:
> Hard links only work across the same filesystem. Do we want to enforce that the kernel image and the jail directory are on the same filesystem?

Hm, I am okay with copying, but as long as the files, kernel image and root drive, are read only this should also be fine. What's your guys' preference, @samuelkarp & @nmeyerhans?

Reply:
Let's copy for now. If we decide that the disk usage an additional IO activity are prohibitively expensive, we can investigate alternatives. Copy is certainly safest from an isolation point of view.